snuba/clickhouse/columns.py

```diff
@@ -15,11 +26,11 @@ def __init__(self, name: str, type: ColumnType) -> None:
     def __repr__(self) -> str:
         return "Column({}, {})".format(repr(self.name), repr(self.type))
 
-    def __eq__(self, other):
+    def __eq__(self, other: object) -> bool:
         return (
             self.__class__ == other.__class__
-            and self.name == other.name
-            and self.type == other.type
+            and self.name == cast(Column, other).name
+            and self.type == cast(Column, other).type
```

Comment:
Wouldn't an `isinstance` check solve the same issue here without the casting? Is there a reason this needs to be an exact match on `__class__` rather than a subclass check?

Reply:
In this use case using isinstance would still be functionally correct, Though that would make `==` return true if we compared an object of class A with an object of a subclass of A that happen to still have the same common attributes. I think this should return False, thus checking they are exactly the same class. What do you think?

Reply:
> I think this should return False, thus checking they are exactly the same class.

I think this is generally a bad practice in object oriented design, but I am not familiar enough with the specific implementation here to have a strong opinion about this individual case since I'm not sure it followed good design principles to begin with

Reply:
In this specific use case it does not matter, since there are no concrete classes extending from concrete classes.
In general my problem in using the isinstance in equality is that, while it is cleaner from an OO design standpoint, it makes equality not symmetric, which I would consider quite confusing and unexpected:

```
@dataclass
class A:
    attr: int

     def __eq__(self, other) -> bool:
           return isinstance(other, self.__class__) and self.attr == other.attr

@dataclass
class B(A):
     attr2: str

     def __eq__(self, other) -> bool:
           return isinstance(other, self.__class__) and self.attr == other.attr and self.attr2 == other.attr2

def compare(c1: A, c2:A):
      assert c1 == c2      

a = A(attr=1)
b = B(attr=1, attr2="str")

compare(a, b)  #ok . 
compare(b, a) .  # fails
```

Reply:
Damn, no, ignore what I said. I got __eq__ wrong. It would always be called on the subclass so it would return False in both cases. So equality does not become non symmetric when using isinstance. Ignore my comment.

Reply:
I guess neither of these are going to be particularly good options because one since maintaining one property is going to sacrifice the other. It probably makes sense to just do what `dataclass` does as precedent (keeping it the way it is now.)

In the future we should probably make these dataclasses (I think we discussed this a long time ago) so that we don't have to make these kinds of decisions and can just lean on a single consistent implementation unless we need to intentionally differ from that. (Please don't do that as part of this change though. Also that will affect https://github.com/getsentry/snuba/pull/835#discussion_r391161694 as noted)

Reply:
> If the operands are of different types, and right operand’s type is a direct or indirect subclass of the left operand’s type, the reflected method of the right operand has priority, otherwise the left operand’s method has priority.

Huh. I did not know this: https://docs.python.org/3/reference/datamodel.html#object.__eq__

I guess wrapping this thread up from my perspective: I don't have a strong preference about the implementation details here after all other than we should probably just move towards making these dataclasses and whatever is implemented here should be consistent with that future implemementation.

Reply:
Actually the isinstance is fine.
TIL that my case above is wrong. Python does this
a == b . ->  b.__eq__(a) .  being b subclass of A
b == a . -> b.__eq__(a).